---
title: "Java 8 Lambda 람다 #2"
excerpt: "Java 8에 추가된 Lambda에 대해 알아보자"
categories: 
  - Java
tags: 
  - Java
  - Lambda
last_modified_at: 2018-11-03T12:00:00+09:00
toc: true
toc_sticky: true
---

## Intro
Java 8에 새로 도입된 Lambda에 대해서 알아보자.  
[Java 8 Lambda 강좌 - Java Brains](https://javabrains.io/courses/java_lambdabasics/ "Java 8 Lambda 강좌 Link")  
[연습예제 github](https://github.com/moregorenine/study/tree/master/core-java/ "연습예제 github Link")

## 람다와 함께 지원되는 functional interfaces
별도의 인터페이스 생성없이 람다표현식을 쉽고 편리하게 사용가능하도록   
Java 8 에서 여러 케이스의 functional interface들을 기본 지원해주는데요.  

개인적으로 MVC에서 사용하는 interface를 이용한 service 구현 패턴을 정말 좋아합니다.  
interface를 service를 구현하면 해당 service가 제공하는 기능들을 일목요연하게 정리된 형태로 보기가 쉽습니다.  
service변경, 대체, 추가등으로 부터 자유로운 범용성을 제공해줍니다.  
해당 서비스들을 사용하는 다른 코드들에 대한 추적도 용이해서 관리적인 면에서도 상당한 매리트를 느끼고 있습니다.  

이런 관점에서 람다식을 보면 관리 추적이 어려워 보이는 람다가 불편해보이기도 하는데요.  
단순히 기능으로 편하게 사용할 함수라던지 Thread에서 Runnable 이라던지 분명한 장점이 있으니 효율적인 사용을 위해선 많은 고민이 필요할 것 같습니다.  

Java 8 에서 지원하는 functional interfaces 와 관련해 내용이 잘 정리된 사이트가 있어 link를 아래 첨부할께요.  

[난파선 프로그래밍 블로그 - 표류중인 블로그](http://napasun-programming.tistory.com/25 "자바에서 제공하는 함수형 인터페이스")  

## 람다식에서의 Exception 처리
람다식에서 Exception 처리는 어떻게 할까요?  
ExeptionExample.java 에서 람다식 호출부분에 Exception을 발생하는 코드를 작성합니다.

### Exception 발생 코드
- ExeptionExample.java
{% highlight java linenos %}
public static void main(String[] args) {
	int[] sumeNumbers = { 1, 2, 3, 4, 5 };
	int key = 2;

	process(sumeNumbers, key, (v, k) -> System.out.println(v / k));
}

private static void process(int[] sumeNumbers, int key, BiConsumer<Integer, Integer> consumer) {
	for (int i : sumeNumbers) {
		consumer.accept(i, key);
	}

}
{% endhighlight %}
- 결과값
```
Exception in thread "main" java.lang.ArithmeticException: / by zero
	at com.github.moregorenine.lambda.unit2.ExeptionExample.lambda$0(ExeptionExample.java:11)
	at com.github.moregorenine.lambda.unit2.ExeptionExample.process(ExeptionExample.java:16)
	at com.github.moregorenine.lambda.unit2.ExeptionExample.main(ExeptionExample.java:11)
```
위와 같이 5line 람다식 실행 중 Exception이 발생했을 때 람다에서는 어떻게 처리해야 할까요?

### Exception 처리 #1 (추천하지 않는 방식)
- ExeptionHandlingExample1.java
{% highlight java linenos %}
public static void main(String[] args) {
	int[] sumeNumbers = { 1, 2, 3, 4, 5 };
	int key = 0;

	process(sumeNumbers, key, (v, k) -> System.out.println(v / k));
}

private static void process(int[] sumeNumbers, int key, BiConsumer<Integer, Integer> consumer) {
	for (int i : sumeNumbers) {
		try {
			consumer.accept(i, key);
		} catch (ArithmeticException e) {
			System.out.println("ArithmeticException 발생...");
		}
	}
}
{% endhighlight %}
- 결과값
```
ArithmeticException 발생...
ArithmeticException 발생...
ArithmeticException 발생...
ArithmeticException 발생...
ArithmeticException 발생...
```
이와 같은 처리는 방식은 해당 process를 사용하는 람다식이 늘어날 경우  
각각의 람다식에 따른 예외처리를 지정할 수 없는 문제가 있습니다.  
그리고 람다식과 Exception 처리 로직이 다른 Class로 분리되었을 경우  
확인하기 불편한 단점이 있습니다.  
그럼 다른 처리 방식을 보시죠.  

### Exception 처리 #2
- ExeptionHandlingExample2.java
{% highlight java linenos %}
public static void main(String[] args) {
	int[] sumeNumbers = { 1, 2, 3, 4, 5 };
	int key = 0;

	process(sumeNumbers, key, (v, k) -> {
		try {
			System.out.println(v / k);
		} catch (Exception e) {
			System.out.println("ArithmeticException 발생...");
		}
	});
}

private static void process(int[] sumeNumbers, int key, BiConsumer<Integer, Integer> consumer) {
	for (int i : sumeNumbers) {
		consumer.accept(i, key);
	}
}
- 결과값
```
ArithmeticException 발생...
ArithmeticException 발생...
ArithmeticException 발생...
ArithmeticException 발생...
ArithmeticException 발생...
```
{% endhighlight %}
## Related Posts
Java 8 Lambda 람다 #1 [link](https://moregorenine.github.io/java/lambda-1/ "Java 8 Lambda 람다 #1")